cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1853,70 +1926,247 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
       r,
-      (_stats_granularity == statistics_freq::STATISTICS_PAGE) ? page_stats.data() : nullptr,
-      (_stats_granularity != statistics_freq::STATISTICS_NONE) ? page_stats.data() + num_pages
-                                                               : nullptr,
-      (_stats_granularity == statistics_freq::STATISTICS_COLUMN) ? page_stats.data() : nullptr);
+      (stats_granularity == statistics_freq::STATISTICS_PAGE) ? page_stats.data() : nullptr,
+      (stats_granularity != statistics_freq::STATISTICS_NONE) ? page_stats.data() + num_pages
+                                                              : nullptr,
+      (stats_granularity == statistics_freq::STATISTICS_COLUMN) ? page_stats.data() : nullptr,
+      compression,
+      column_index_truncate_length,
+      stream);
+
+    bool need_sync{false};
 
-    std::vector<std::future<void>> write_tasks;
     for (; r < rnext; r++) {
       int p           = rg_to_part[r];
       int global_r    = global_rowgroup_base[p] + r - first_rg_in_part[p];
-      auto& row_group = _agg_meta->file(p).row_groups[global_r];
+      auto& row_group = agg_meta->file(p).row_groups[global_r];
+
       for (auto i = 0; i < num_columns; i++) {
-        gpu::EncColumnChunk& ck = chunks[r][i];
+        auto const& ck          = chunks[r][i];
+        auto const dev_bfr      = ck.is_compressed ? ck.compressed_bfr : ck.uncompressed_bfr;
         auto& column_chunk_meta = row_group.columns[i].meta_data;
-        uint8_t* dev_bfr;
-        if (ck.is_compressed) {
-          column_chunk_meta.codec = _compression;
-          dev_bfr                 = ck.compressed_bfr;
-        } else {
-          dev_bfr = ck.uncompressed_bfr;
+
+        if (ck.is_compressed) { column_chunk_meta.codec = compression; }
+
+        if (ck.ck_stat_size != 0) {
+          column_chunk_meta.statistics_blob.resize(ck.ck_stat_size);
+          CUDF_CUDA_TRY(cudaMemcpyAsync(column_chunk_meta.statistics_blob.data(),
+                                        dev_bfr,
+                                        ck.ck_stat_size,
+                                        cudaMemcpyDefault,
+                                        stream.value()));
+          need_sync = true;
         }
 
+        row_group.total_byte_size += ck.compressed_size;
+        column_chunk_meta.total_uncompressed_size = ck.bfr_size;
+        column_chunk_meta.total_compressed_size   = ck.compressed_size;
+      }
+    }
+
+    // Sync before calling the next `encode_pages` which may alter the stats data.
+    if (need_sync) { stream.synchronize(); }
+  }
+
+  auto write_buff = pinned_buffer<uint8_t>{[](size_t size) {
+                                             uint8_t* ptr = nullptr;
+                                             CUDF_CUDA_TRY(cudaMallocHost(&ptr, size));
+                                             return ptr;
+                                           }(max_chunk_bfr_size),
+                                           cudaFreeHost};
```

Comment:
Maybe a good time to replace pinned_buffer with a `thrust::host_vector<uint8_t, rmm::mr::pinned_memory_resource>` ?

Reply:
Good point. But I also want to change that for ORC writer too. Will do it in a separate PR.